### PR TITLE
feat(pt): support mixed-nloc LMDB batching

### PR DESCRIPTION
**Pull request by littlepeachs**
# Add mixed-nloc LMDB batching for PyTorch training

## Summary

This PR adds optional `mixed_batch` support to the PyTorch LMDB training and validation data path.

When `mixed_batch` is enabled, one LMDB batch may contain frames with different atom counts / `nloc` values. Atom-wise fields are flattened across the batch, frame ownership is tracked with `batch` and `ptr`, and the model runs through a flat forward path without padding.

The default behavior is unchanged. `mixed_batch` is `false` by default, so LMDB batches still use the existing same-`nloc` sampler unless the option is explicitly enabled.

## Motivation

LMDB batches previously required all frames in a batch to have the same `nloc`. This can make batching inefficient for datasets that mix systems with different atom counts. Mixed batching lets those frames train together while keeping atom indexing explicit in a flattened graph layout.

## Main Changes

- Add `mixed_batch` to PyTorch LMDB `training_data` and `validation_data`.
- Keep `mix_batch` as an alias for the same option.
- In `mixed_batch=true` mode, interpret `batch_size` as the number of frames/systems in one mixed batch.
- Add a mixed LMDB collate path that concatenates atom-wise fields, stacks frame-wise fields, and emits `batch` and `ptr` for frame ownership.
- Precompute flat graph metadata and indices during collation for the DPA3/RepFlow path.
- Add flat forward support through the training wrapper, energy model, atomic model, descriptors, fitting net, and energy loss.
- Keep atom-wise outputs flat and reduce frame-wise quantities using the global atom-to-frame mapping.

## Usage

```json
{
  "training": {
    "training_data": {
      "systems": "path/to/train.lmdb",
      "batch_size": 8,
      "mixed_batch": true
    },
    "validation_data": {
      "systems": "path/to/valid.lmdb",
      "batch_size": 8,
      "mixed_batch": true
    }
  }
}
```

With `mixed_batch=true`, `batch_size` controls how many frames/systems are merged into one flattened batch.

With `mixed_batch=false`, the existing same-`nloc` LMDB batching behavior is preserved.

## Tests

- Add unit tests for mixed-nloc LMDB collation.
- Cover flattened `batch` / `ptr` layout and flat graph precomputation.
- Cover the `mix_batch` config alias.
- Add a DPA3 LMDB mixed-batch example config.
- Add `test_mixed_batch.sh` for end-to-end training coverage.

## Scope

- This change applies to the PyTorch LMDB data path.
- The flat graph path is currently used by the DPA3/RepFlow descriptor flow.

## Summary by CodeRabbit

* **New Features**
  * Mixed-batch LMDB training: flat atom-wise collation, optional precomputed flat-graph inputs, and model paths to consume them.
  * Trainer and model wrapper accept flat-graph inputs for mixed-nloc workloads.
  * Deterministic default initialization seeds for model components.

* **Improvements**
  * Per-frame normalization for energy and virial losses in mixed batches.

* **Documentation**
  * Added mixed-batch LMDB batching and flat-graph preprocessing guide.

* **Tests**
  * Added unit/integration tests, sample configs, and a runnable mixed-batch training script.

[Review Change Stack](https://app.coderabbit.ai/change-stack/deepmodeling/deepmd-kit/pull/5439?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: .gitignore
Comment (iProzd) [inline .gitignore:1]: This modification should be removed.

File: test_mixed_batch.sh
Comment (iProzd) [inline test_mixed_batch.sh:1]: This modification should be removed.

File: test_mptraj/lmdb_baseline.json
Comment (iProzd) [inline test_mptraj/lmdb_baseline.json:1]: This modification should be removed.

File: test_mptraj/lmdb_mixed_batch.json
Comment (iProzd) [inline test_mptraj/lmdb_mixed_batch.json:1]: This modification should be removed.

File: doc/development/lmdb-mixed-system-batching.md
Comment (iProzd) [inline doc/development/lmdb-mixed-system-batching.md:1]: Doc should be written in english, and placed as a **feature introduction** instead of a development doc.

**Review by iProzd — CHANGES_REQUESTED** [review]
1. Implementation in other backend (at least dpmodel backend) should be added.

2. Tests for new forward method, such as universal tests in `source/tests/universal`, end-to-end training in `source/tests/pt/test_training.py` and consistent tests in `source/tests/consistent` should be added.

File: .gitignore
Comment (littlepeachs) [reply .gitignore:1]: Addressed. The `.gitignore` change has been removed from the PR and restored to match the upstream master baseline.

I kept my local-only ignore patterns in `.git/info/exclude`, so they do not affect the PR.

File: test_mixed_batch.sh
Comment (littlepeachs) [reply test_mixed_batch.sh:1]: Addressed. `test_mixed_batch.sh` has been removed from the PR diff. It was only a local/debug helper and is no longer tracked in this PR.

File: test_mptraj/lmdb_baseline.json
Comment (littlepeachs) [reply test_mptraj/lmdb_baseline.json:1]: Addressed. `test_mptraj/lmdb_baseline.json` has been removed from the PR diff. It was a local/debug config and is no longer tracked in this PR.

File: test_mptraj/lmdb_mixed_batch.json
Comment (littlepeachs) [reply test_mptraj/lmdb_mixed_batch.json:1]: Addressed. `test_mptraj/lmdb_mixed_batch.json` has been removed from the PR diff. It was a local/debug config and is no longer tracked in this PR.

File: doc/development/lmdb-mixed-system-batching.md
Comment (littlepeachs) [reply doc/development/lmdb-mixed-system-batching.md:1]: Addressed for this PR. I removed `doc/development/lmdb-mixed-system-batching.md` from the PR diff, since the current file was a development/debug note and not the requested English feature introduction.

I will add a proper user-facing English feature introduction separately if documentation is still required.

**Comment by littlepeachs** [comment]
@iProzd Thanks for the review. I addressed the requested backend implementation and tests.

1. Added dpmodel backend support:
   - `deepmd/dpmodel/model/ener_model.py`
   - `source/tests/universal/dpmodel/model/test_mixed_batch_flat_forward.py`

`EnergyModel.call` in dpmodel now accepts `batch` / `ptr` and the flat graph argument set. When `batch` / `ptr` are provided, it routes to a new `call_flat` path. The dpmodel implementation splits the flat mixed-nloc batch by `ptr`, reuses the existing regular per-frame dpmodel call path, and merges translated outputs back into the mixed-batch flat layout. Hessian is explicitly rejected in this flat path to avoid silently returning incorrect results.

2. Added/expanded tests for the new forward/data path:
   - `source/tests/universal/dpmodel/model/test_mixed_batch_flat_forward.py`
   - `source/tests/universal/pt/model/test_mixed_batch_flat_forward.py`
   - `source/tests/consistent/test_lmdb_data.py`
   - `source/tests/pt/test_lmdb_dataloader.py`
   - `source/tests/pt/test_training.py`

3. Removed local/debug files from the PR as requested:
   - `.gitignore`
   - `test_mixed_batch.sh`
   - `test_mptraj/lmdb_baseline.json`
   - `test_mptraj/lmdb_mixed_batch.json`
   - `doc/development/lmdb-mixed-system-batching.md`

The latest CI on head `23c26335` finished with 0 failures.

**Comment by iProzd** [comment]
Resolved in #5962
